### PR TITLE
James Construction (Connectivity)

### PR DESCRIPTION
**Pull request by kangrongji**
This PR contains the connectivity results of James construction. It also contains necessary facts about direct colimits and pushout-product. I think the general theory of James has been fully formalized. @aljungstrom @mortberg

### CODE REVIEW

**Comment by mortberg** [comment]
Very nice! It would be very interesting to see if we can use this to get the more general result in Guillaume's thesis formalized... @aljungstrom  What do you think?

(As you might have noticed I've been very quiet here lately. I'm extremely busy with teaching at the moment and will be for a few more weeks, so won't have time to review any PRs for a while longer).

**Comment by kangrongji** [comment]
I hope this could be help:)

**Comment by aljungstrom** [comment]
Very nice! @mortberg I don't think the more general statement in Brunerie's thesis is that interesting (I think we get the remaining homotopy groups by Freudenthal). But there's definitely more stuff we can explore now that we have James!

**Comment by kangrongji** [comment]
Any update on your "general result"? @aljungstrom

**Comment by aljungstrom** [comment]
> Any update on your "general result"? @aljungstrom

Oh no unfortunately I had nothing great in mind. I just confused Anders' question with another one we had discussed earlier that day (which was nothing too exciting - just an application of Freudenthal).

**Comment by ecavallo** [comment]
What are the dependencies between the James PRs? Should we merge #718 and #722 before this one?

**Comment by kangrongji** [comment]
> What are the dependencies between the James PRs? Should we merge #718 and #722 before this one?

Yes, each one depends on the previous one.

File: Cubical/HITs/SequentialColimit/Properties.agda [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        liftingData sec .finl  = lifting _ sec
+        liftingData sec .fpush = liftingPath _ sec
+
+        hasSectionInl∘ : hasSection (λ (sec : (x : Lim→ X) → Y x .fst) → sec ∘ inl {n = n})
+        hasSectionInl∘ .fst sec = elimShift  _ _   (liftingData sec)
+        hasSectionInl∘ .snd sec = elimShiftβ _ _ _ (liftingData sec)
+
+    -- Connectivity of inclusion map
+
+    isConnInl∞ : isConnectedFun d (inl∞ n)
```
Comment (ecavallo) [inline Cubical/HITs/SequentialColimit/Properties.agda:236]: isConn -> isConnected

File: Cubical/HITs/James/Inductive/PushoutFormula.agda [hunk 2]
```diff
@@ -166,3 +174,98 @@ module _
...
+    (subst (λ d → isConnectedFun d (incl {X∙ = X∙} {n = k + m}))
+      (nat-path n m k) (isConnIncl n conn (k + m)))
+
+  private
+    inl∞ : (n : ℕ) → 𝕁ames n → 𝕁ames∞
+    inl∞ _ = inl
+
+  isConnInl : (n : ℕ) → isConnected (1 + n) X
+    → (m : ℕ) → isConnectedFun ((1 + m) · n) (inl∞ m)
+  isConnInl n conn m = isConnInl∞ _ _ _ (isConnIncl>n _ conn _)
```
Comment (ecavallo) [inline Cubical/HITs/James/Inductive/PushoutFormula.agda:271]: isConn -> isConnected

**Review by ecavallo — CHANGES_REQUESTED** [review]
Just a superficial request: I want to always use `isConnected[...]` in names of connectedness theorems. (But it's fine if this isn't followed in private definitions.)

Nice idea to introduce a shifted elimination principle for the sequential colimit. I had so much trouble with associativity trying to prove things about these a long time ago...

File: Cubical/HITs/Pushout/PushoutProduct.agda [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        ; (j = i1) → ϕ i1 x .snd y i })
+      (ϕ j x .snd y i)
+
+  lifting : hasSection (λ(s : (x : A × B) → P x .fst) → s ∘ (f ×̂ g))
+  lifting .fst sec = ext sec
+  lifting .snd sec i x = ext-path sec x i
+
+-- The connectivity of pushout-product
+
+isConn×̂ : {m n : ℕ}{f : A → B}{g : X → Y}
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/HITs/Pushout/PushoutProduct.agda:103]: done

File: Cubical/HITs/SequentialColimit/Properties.agda (hunk 1 above)
Comment (kangrongji) [reply Cubical/HITs/SequentialColimit/Properties.agda:236]: done

File: Cubical/HITs/James/Inductive/PushoutFormula.agda (hunk 2 above)
Comment (kangrongji) [reply Cubical/HITs/James/Inductive/PushoutFormula.agda:271]: done

File: Cubical/HITs/James/Inductive.agda [hunk 4]
```diff
@@ -80,8 +93,29 @@ module JamesInd
...
+  module _
+    (d : ℕ)(conn : isConnected (1 + d) X) where
+
+    -- Warning:
+    -- The connectivity is shifted by 2 from the convention of usual homotopy theory.
+
+    -- If X is (d+1)-connected, the transition incl : 𝕁 n → 𝕁 (n+1) will be (n+1)d-connected.
+
+    inclConn : (n : ℕ) → isConnectedFun ((1 + n) · d) (incl {n = n})
+    inclConn = isConnIncl X∙ d conn
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/HITs/James/Inductive.agda:113]: done

**Comment by kangrongji** [comment]
Yup, those who parametrized over natural numbers are just painful and messy.

Btw, I haven't noticed but after you guys clean up all the PRs I realize there has been a WIP on direct limit #592 . It seems the only hole left there is just the shifted eliminators I show here. So it'd better ping @felixwellen .

**Review by ecavallo — APPROVED** [review]
Thanks!

**Pull request merged**